Raise exceptions by default; opt to silence in message listener task

fselmo opened the pull request:
### What was wrong?

It is often best to fail hard and fast and be able to silence exceptions rather than let exceptions possibly go unnoticed. We were defaulting to silencing exceptions in the message listener task for `WebsocketProviderV2`. This hasn't been released yet and so this change just aims to tweak the configuration and opt to silence exceptions that are raised by default.

### How was it fixed?

- `raise_listener_task_exceptions` -> `silence_listener_task_exceptions`
- Modify tests accordingly
- Close all current tasks in the loop before raising the exception

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![20240123_141138](https://github.com/ethereum/web3.py/assets/3532824/9f72472d-341c-4699-a7a4-aa078202fbf2)

fselmo committed:
Raise exceptions by default; opt to silence in message listener task

fselmo committed:
newsfragment for #3202

pacrob commented:
Update looks good, but I don't see any reference to this option in docs. Is that coming separately, or should be added here?

fselmo committed:
Update documentation for ``WebsocketProviderV2`` + minor refactor.

- Remove the ``endpoint_uri`` argument from the ``PersistentConenctionProvider`` base class, opting for defining it in the inheriting classes instead.
- Update the documentation around the ``WebsocketProviderV2`` class to reflect the current state + document ``PersistentConenctionProvider`` under a new section.

pacrob reviewed (approved):
lgtm!

fselmo commented:
@pacrob I added a more extensive update of the docs here since it was stale. Thanks for the poke. Take a look again and lmk if anything looks off. I also took the `endpoint_uri` out of the base ``PersistentConnectionProvider`` class, opting to only keep it in the inheriting classes which are the actual provider classes.

pacrob reviewed (approved):
lgtm!
